src/analysis/calculus/iterated_deriv.lean

```diff
@@ -0,0 +1,296 @@
+      continuous_linear_equiv.self_comp_symm],
+  refl
+end
+
+/-- The `n`-th Fréchet derivative applied to a vector `(m 0, ..., m (n-1))` is the derivative
+multiplied by the product of the `m i`s. -/
+lemma iterated_fderiv_apply_eq_iterated_deriv_mul_prod {m : (fin n) → 𝕜} :
+  (iterated_fderiv 𝕜 n f x : ((fin n) → 𝕜) → F) m = finset.univ.prod m • iterated_deriv n f x :=
+by { rw [iterated_deriv_eq_iterated_fderiv, ← continuous_multilinear_map.map_smul_univ], simp }
+
+@[simp] lemma iterated_deriv_zero :
+  iterated_deriv 0 f = f :=
+by { ext x, simp [iterated_deriv] }
+
+@[simp] lemma iterated_deriv_one :
+  iterated_deriv 1 f = deriv f :=
+by { ext x, simp [iterated_deriv], refl }
+
+/-- The property of being `C^n`, initially defined in terms of the Fréchet derivative, can be
+reformulated in terms of the one-dimensional derivative. -/
+lemma times_cont_diff_iff_iterated_deriv {n : with_top ℕ} :
+  times_cont_diff 𝕜 n f ↔
+(∀m:ℕ, (m : with_top ℕ) ≤ n → continuous (iterated_deriv m f))
+∧ (∀m:ℕ, (m : with_top ℕ) < n → differentiable 𝕜 (iterated_deriv m f)) :=
+by simp only [times_cont_diff_iff_continuous_differentiable, iterated_fderiv_eq_equiv_comp,
+  continuous_linear_equiv.comp_continuous_iff,
+  continuous_linear_equiv.comp_differentiable_iff]
+
+/-- To check that a function is `n` times continuously differentiable, it suffices to check that its
+first `n` derivatives are differentiable. This is slightly too strong as the condition we
+require on the `n`-th derivative is differentiability instead of continuity, but it has the
+advantage of avoiding the discussion of continuity in the proof (and for `n = ∞` this is optimal).
+-/
+lemma times_cont_diff_of_differentiable_iterated_deriv {n : with_top ℕ}
+  (h : ∀(m : ℕ), (m : with_top ℕ) ≤ n → differentiable 𝕜 (iterated_deriv m f)) :
+  times_cont_diff 𝕜 n f :=
+times_cont_diff_iff_iterated_deriv.2
+  ⟨λ m hm, (h m hm).continuous, λ m hm, (h m (le_of_lt hm))⟩
+
+lemma times_cont_diff.continuous_iterated_deriv {n : with_top ℕ} (m : ℕ)
+  (h : times_cont_diff 𝕜 n f) (hmn : (m : with_top ℕ) ≤ n) :
+  continuous (iterated_deriv m f) :=
+(times_cont_diff_iff_iterated_deriv.1 h).1 m hmn
+
+lemma times_cont_diff.differentiable_iterated_deriv {n : with_top ℕ} (m : ℕ)
+  (h : times_cont_diff 𝕜 n f) (hmn : (m : with_top ℕ) < n) :
+  differentiable 𝕜 (iterated_deriv m f) :=
+(times_cont_diff_iff_iterated_deriv.1 h).2 m hmn
+
+/-- The `n+1`-th iterated derivative can be obtained by differentiating the `n`-th
+iterated derivative. -/
+lemma iterated_deriv_succ : iterated_deriv (n + 1) f = deriv (iterated_deriv n f) :=
+begin
+  ext x,
+  rw [← iterated_deriv_within_univ, ← iterated_deriv_within_univ, ← deriv_within_univ],
+  exact iterated_deriv_within_succ unique_diff_within_at_univ,
+end
+
+/-- The `n`-th iterated derivative can be obtained by iterating `n` times the
+differentiation operation. -/
+lemma iterated_deriv_eq_iterate : iterated_deriv n f = nat.iterate deriv n f :=
```

Comment:
Why not use `deriv^[n] f` as the main API for iterated 1-dimensional derivatives?

Reply:
I do agree that I like the notation a lot...

Reply:
`^[n]` is a notation for `nat.iterate`. My question is why @sgouezel chose to define `iterated_deriv` instead of using `nat.iterate deriv`. This way `deriv^[0] f = f` and `deriv^[1] f = deriv f` would be definitional equalities.

Reply:
I think Mario would tell you that you shouldn't care too much about the defeqs in this case. But nevertheless I think that `nat.iterate` is a natural candidate to apply here, and in addition it unlocks the nice notation.

Reply:
In domains, there is a choice involved in the definition of the derivative. If you defined the iterated derivative within a set as `nat.iterate (deriv_within s)`, then the choice might be different than the one for the `n`-th Fréchet derivative, and you might get unrelated objects. So, it is better to define it starting from the `n`-th Fréchet derivative in the domain, to ensure consistency everywhere. 

For the derivative in the whole space, the choice is unique, so in the end you would get the same object if you defined it with `nat.iterate deriv`. Still, to get a uniform interface and uniform statements within sets and on the space, I rather followed the same definition as above and started from the Fréchet derivative.

Anyway, for complicated objects like that, defeq is not really relevant, like Johan said.